### DISCUSSION TITLE
Escaping string key and values while dumping

### DISCUSSION THREAD

**Discussion by persona94** [nlohmann/json#4249] [Q&A] [answered]
If I have a json

```
json j = json::parse(R"({"a": 122341, "b": "testing"})");
```

How do I convert `j` to a fully escaped std::string?

Expected

```
std::string s = json::escaped_string(j)
std::cout << j << std::endl;
```

With the expected output of 

```
{\"a\":122341,\"b\":\"testing\"}
```

I can write my own function that escapes the quotes but wondering if there was some function native to the lib that did that already

**Comment by nlohmann** [comment]
Use dump: https://json.nlohmann.me/api/basic_json/#serialization-dumping

**Reply by persona94** [reply]
I started with that actually and it didn't work

```
json j = json::parse(R"({"a": 122341, "b": "testing"})");
std::cout << j.dump() << std::endl;
```

Output

```
{"a":122341,"b":"testing"}
```

**Reply by nlohmann** [reply]
Then try json(j.dump()).get<std::string>()

**Reply by persona94** [reply]
Same results

```
json j = json::parse(R"({"a": 122341, "b": "testing"})");
std::cout << json(j.dump()).get<std::string>() << std::endl;
```

Output

```
{"a":122341,"b":"testing"}
```

**Reply by nlohmann** [reply, marked as answer]
Try

std::cout << json(j.dump()).dump() << std::endl;

May I ask why you need such an escaped string?

**Reply by persona94** [reply]
Thank you for the help! That did work. I'll have to remove the leading and trailing dquotes but that's manageable

```
"{\"a\":122341,\"b\":\"testing\"}"
```

In terms of why, I am trying to use the spdlog library to generate a json log. The message is sometimes in JSON format.

So if I try

```
nlohmann::json j;
  
j["k1"] = "v1";
j["k2"] = 2;
LOG_ERROR(j.dump());
```

The output I get is 

```{"@timestamp":"2023-12-27T15:46:41.636829Z","message":"{"k1":"v1","k2":2}","level":"error","tid":25,"tuid":11,"source":"main.cpp","function":"testPost","line":"23"}```

If you take this output JSON as a whole then it's an invalid JSON. The only way to make it "valid" would be to escape every dquote within ```{"k1":"v1","k2":2}``` so that the final log looks like

```{"@timestamp":"2023-12-27T15:46:41.636829Z","message":"{`"k1`":`"v1`",`"k2`":2}","level":"error","tid":25,"tuid":11,"source":"main.cpp","function":"testPost","line":"23"}```

This is due to the log needing to go on to elastic search. The log library does not support JSON as such so it's up to me to provide a JSON string that's escaped per my requirements.

Until you suggested the double dump I had written a function that use string streams and std::quote. It got me an escaped string but it made me wonder if there was something in the json lib itself that could directly get me an escaped string. It would be very useful to have a param like `json.dump(bool escapeQuotes)` where `escapeQuotes` is false by default which satisfies 99% of the use cases but can help in situations like mine. I'm sure I'm not the only one who has had to escape dquotes for usage in logs or somewhere online. Regardless, thanks for the suggestions.
